### PR TITLE
More DragonBattle API

### PR DESCRIPTION
**Pull request by Machine-Maker**
Adds methods for getting different end crystals directly related to the DragonBattle. Also adds methods for dealing with gateways.

### CODE REVIEW

**Comment by Machine-Maker** [comment]
Rebased and tested for 1.19.3

**Review by Owen1212055 — APPROVED** [review]

File: patches/server/0951-More-DragonBattle-API.patch [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
++        return java.util.Collections.unmodifiableList(enderCrystals);
++    }
++
++    @Override
++    public java.util.List<org.bukkit.entity.EnderCrystal> getHealingCrystals() {
++        final java.util.List<org.bukkit.entity.EnderCrystal> enderCrystals = new java.util.ArrayList<>();
++        for (final net.minecraft.world.entity.boss.enderdragon.EndCrystal endCrystal : this.handle.getSpikeCrystals()) {
++            if (!endCrystal.isRemoved() && endCrystal.isAlive() && endCrystal.valid) {
++                enderCrystals.add(((org.bukkit.entity.EnderCrystal) endCrystal.getBukkitEntity()));
++            }
```
Comment (Owen1212055) [inline patches/server/0951-More-DragonBattle-API.patch]: All these alive checks seem redundant?

Comment (Machine-Maker) [reply patches/server/0951-More-DragonBattle-API.patch]: Vanilla doesn’t actually remove the entity from the list when it dies as far as I can tell. So it just has a list of non-existent entities. That’s there to just make sure they don’t sneak into the api.

File: patches/api/0423-More-DragonBattle-API.patch [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
++     */
++    boolean spawnNewGateway();
++
++    /**
++     * Spawns a new end gateway at the specified position. This will
++     * spawn regardless of the number of gateways already present.
++     *
++     * @param position position for the new gateway
++     */
++    void spawnNewGateway(@NotNull io.papermc.paper.math.Position position);
```
Comment (kennytv) [inline patches/api/0423-More-DragonBattle-API.patch]: just `spawnGateway` or `placeGateway` or `createGateway` or something

File: patches/server/0951-More-DragonBattle-API.patch [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+     private void spawnNewGateway() {
+         if (!this.gateways.isEmpty()) {
+@@ -541,6 +550,15 @@ public class EndDragonFight {
+         }
+ 
+     }
++    // Paper start
++    public List<EndCrystal> getSpikeCrystals() {
++        List<EndCrystal> endCrystals = new java.util.ArrayList<>();
++        for(SpikeFeature.EndSpike endSpike : SpikeFeature.getSpikesForLevel(this.level)) {
```
Comment (kennytv) [inline patches/server/0951-More-DragonBattle-API.patch]: space

**Review by kennytv — APPROVED** [review]

**Pull request merged**